### PR TITLE
dev-libs/date: add 3.0.3 with one more test working

### PR DESCRIPTION
**Pull request by fkobi**
Bug: https://bugs.gentoo.org/940170

---

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

### CODE REVIEW

**Comment by JonasToth** [comment]
thank you for the patch!
i believe the things i commented on could move into a different ebuild-version that closer tracks master. the normal pattern is `9999` that may suffice for now, as there was a new official release? what do you think about that?

**Comment by JonasToth** [comment]
Sorry for beeing so unclear. The ebuild for the tagged Version has elements of a live/snapshoted ebuild in it. I thought it makes sense to create a live ebuild but keep the official Releases free from those elements (the commit Reference and so)

**Comment by eli-schwartz** [comment]
A snapshot ebuild is not the same as a live ebuild -- live ebuilds are a special property indicating that every time you re-run the compile for that ebuild, it potentially produces a different software product based on whatever the most recent state of the upstream codebase is. Usually, this means downloading the source as a git repository and `git pull`ing the latest version each time.

Consequently, live ebuilds:
- have no version
- have no stable checksum
- intentionally violate portage's network sandboxing
- cannot have KEYWORDS=

A snapshot has checksums, it reliably produces the same results every time you build it, it just isn't a tagged version usually because upstream is, well, not so great at releasing new versions. ;) But the point of a snapshot being packaged is because it is potentially desirable to stabilize it.

The point of a live ebuild is because some users like to experiment with the very latest bleeding edge code, and this lets them opt into things like checking if some bug has indeed been fixed 5 minutes after their bug report was closed upstream with "thanks for the report, fixed in commit XXXXX".

If there's a general intention that the proxied maintainer of this package feels it is appropriate to, on a semi-regular basis, package new snapshots? Then yes, it sounds worth it to provide some mechanism in the ebuild to make that a bit easier and require less manual updating each time.

Was the previous snapshot a one-off? Something that is expected to regularly repeat?

**Comment by JonasToth** [comment]
> I authored the latest ebuild. It is a snapshot as the release 3.0.1 was made just before a commit that fixed a bunch of failing tests -- it's basically 3.0.1 + [HowardHinnant/date@155c6b9](https://github.com/HowardHinnant/date/commit/155c6b9e76e462e1d47ea528ca87f366adccdea3).
> 
> Then I thought that this would be a one-off, but looking at the [release history](https://github.com/HowardHinnant/date/releases), [how much effort is put into releases](1) and the fact that upstream [recommends "living at HEAD"](https://github.com/HowardHinnant/date/issues/828#issuecomment-2184283807) I think shipping snapshots is the way to go here.

I agree, that the packaging is not the normal case for this library. For the releases 3.0.2 and 3.0.3 I would still like classical ebuilds as they are classical releases.

And "living at head" matches perfectly to a live ebuild, so that should be done I guess.

The pro for snapshoted ebuilds is, that they can get stabilized. It is expected, that no frequent releases will occur in the next months, so an inbetween release makes sense. I am not opposed to this! I think, the classical ebuilds don't need the infrastructure for the commit selection and so on. Thats what I was targeting in my comments.

@fkobi TBH, you have more knowledge than I do. I am not a frequent contributor and kinda lost time for heavy open source work :/

**Review by JonasToth — APPROVED** [review]
Thanks, looks good!

File: dev-libs/date/date-3.0.2.ebuild [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# Copyright 1999-2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+inherit cmake
+
+# maintainer recommends to "live at head" so we can ship snapshots
```
[the remark this answers is not in the document]
Comment (JonasToth) [reply dev-libs/date/date-3.0.2.ebuild]: is resolved.

File: dev-libs/date/date-3.0.2.ebuild [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+# COMMIT="155c6b9e76e462e1d47ea528ca87f366adccdea3" # 2024.09.13
+
+if [[ -n $COMMIT ]]; then
+	SRC_URI="https://github.com/HowardHinnant/date/archive/${COMMIT}.tar.gz -> ${P}.gh.tar.gz"
```
[the remark this answers is not in the document]
Comment (JonasToth) [reply dev-libs/date/date-3.0.2.ebuild]: is resolved

**Comment by JonasToth** [comment]
> For me doing snapshot ebuilds is simpler than patching in some commits but I am willing to work with patches if that is what you prefer -- I'll get some exercise then :D

I did see this often in gentoo for various reasons. Maybe a gentoo dev has more insights on what should be prefered. Lets see :)

**Review by JonasToth — APPROVED** [review]

**Comment by thesamesam** [comment]
Aggressively preferring snapshots can be an error if e.g. the next release becomes branched and someone doesn't notice or changes which aren't really necessary to have downstream end up backported via them and upstream didn't intend/expect anyone to do that.

File: dev-libs/date/date-3.0.3.ebuild [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+SLOT="0/${PV}"
+KEYWORDS="~amd64 ~arm64"
+IUSE="only-c-locale test"
+RESTRICT="!test? ( test )"
+
+BDEPEND="test? ( sys-devel/clang )" # tests call clang++
+
+PATCHES=( "$FILESDIR"/${PN}-3.0.3_remove-failing-tests.patch )
+
+src_prepare() {
```
Comment (thesamesam) [inline dev-libs/date/date-3.0.3.ebuild:22]: There's no need for this, but I'll delete it myself in the interest of getting it in.

File: dev-libs/date/date-3.0.3.ebuild [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+LICENSE="MIT"
+SLOT="0/${PV}"
+KEYWORDS="~amd64 ~arm64"
+IUSE="only-c-locale test"
+RESTRICT="!test? ( test )"
+
+BDEPEND="test? ( sys-devel/clang )" # tests call clang++
+
+PATCHES=( "$FILESDIR"/${PN}-3.0.3_remove-failing-tests.patch )
```
Comment (thesamesam) [inline dev-libs/date/date-3.0.3.ebuild:20]:
```suggestion
PATCHES=( "${FILESDIR}"/${PN}-3.0.3_remove-failing-tests.patch )
```

File: dev-libs/date/date-3.0.3.ebuild [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+HOMEPAGE="https://github.com/HowardHinnant/date"
+SRC_URI="https://github.com/HowardHinnant/date/archive/refs/tags/v${PV}.tar.gz -> ${P}.gh.tar.gz"
+
+LICENSE="MIT"
+SLOT="0/${PV}"
+KEYWORDS="~amd64 ~arm64"
+IUSE="only-c-locale test"
+RESTRICT="!test? ( test )"
+
+BDEPEND="test? ( sys-devel/clang )" # tests call clang++
```
Comment (thesamesam) [inline dev-libs/date/date-3.0.3.ebuild:18]: Sounds like that would be worth investigating and fixing...

File: dev-libs/date/files/date-3.0.3_remove-failing-tests.patch [hunk 6]
```diff
@@ -0,0 +1,28 @@
+From: Filip Kobierski <fkobi@pm.me>
```
Comment (thesamesam) [inline dev-libs/date/files/date-3.0.3_remove-failing-tests.patch:1]: But why are we deleting this? Have you investigated why they're failing?

**Comment by JonasToth** [comment]
> With reluctance, I'll merge it with some fixups, but please investigate my comments.

I will look into the issues. I did check the tests definition of upstream and they do support other compilers than clang on first sight.

File: dev-libs/date/files/date-3.0.3_remove-failing-tests.patch (hunk 6 above)
Comment (thesamesam) [reply dev-libs/date/files/date-3.0.3_remove-failing-tests.patch:1]: The ebuild / commit message should reflect that then...

**Comment by JonasToth** [comment]
> > With reluctance, I'll merge it with some fixups, but please investigate my comments.
> 
> I will look into the issues. I did check the tests definition of upstream and they do support other compilers than clang on first sight.

Fkobi was much faster. I think the issues are properly adressed now and I wont do more.

**Pull request closed without merging**